python_bindings/tests/end-to-python-end-test/end_utils.py

```diff
@@ -0,0 +1,110 @@
+from typing import NamedTuple, Tuple
+
+from carcassonne_engine import GameEngine, SerializedGame
+from carcassonne_engine._bindings.elements import MeepleType
+from carcassonne_engine._bindings.feature import Type as FeatureType
+from carcassonne_engine._bindings.side import Side
+from carcassonne_engine.placed_tile import PlacedTile, Position, Tile
+from carcassonne_engine.requests import (
+    GetLegalMovesRequest,
+    MoveWithState,
+    PlayTurnRequest,
+)
+
+
+class TurnParams(NamedTuple):
+    pos: Position
+    tile: Tile
+    meepleType: MeepleType
+    side: Side
+    featureType: FeatureType
+
+
+def find_meeple_in_placed_tile(ptile: PlacedTile):
+    for feature in ptile._go_obj.Features:
+        if feature.Meeple.Type != 0:
+            print(f"sides: {feature.Sides}, feat type: {feature.FeatureType}")
+            return feature.Meeple
+    print("NONE")
+    return None
+
+
+def get_placed_tile(moves: list[MoveWithState], turnParams: TurnParams) -> PlacedTile:
+    for move in moves:
+        # if exact placement
+        if (
+            move.move.to_tile().exact_equals(turnParams.tile)
+            and move.move.position == turnParams.pos
+        ):
+            if turnParams.meepleType == MeepleType.NoneMeeple:
+                meepleExists = False
+                # check if there is no meeple on a tile
+                for feature in move.move._go_obj.Features:
+                    if feature.Meeple.Type != MeepleType.NoneMeeple.value:
+                        meepleExists = True
+                        break
+                if not meepleExists:
+                    return move.move
+            else:
+                # check if meeple in desired position
+                # find_meeple_in_placed_tile(move.move)
+                feature = move.move._go_obj.GetPlacedFeatureAtSide(
+                    sideToCheck=turnParams.side.value,
+                    featureType=turnParams.featureType.value,
+                )
+                if (
+                    feature is not None
+                    and feature.Meeple.Type != turnParams.meepleType.NoneMeeple.value
+                ):
```

Comment:
Python if statements like this one generally shouldn't use parentheses

Reply:
That's not true - the formatting style we selected (and which is enforced by our autoformatter) uses parens to split long expressions into multiple lines.

Reply:
Oh, I didn't know that, sorry